src/main/java/io/opentracing/contrib/global/GlobalTracer.java

```diff
@@ -0,0 +1,150 @@
+package io.opentracing.contrib.global;
+
+import io.opentracing.NoopTracerFactory;
+import io.opentracing.SpanContext;
+import io.opentracing.Tracer;
+import io.opentracing.propagation.Format;
+
+import java.util.Iterator;
+import java.util.ServiceLoader;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * The {@linkplain GlobalTracer} forwards all methods to a single {@link Tracer} implementation that can be
+ * instantiated in one of two ways:
+ * <ol>
+ * <li>Explicitly, by calling {@link #set(Tracer)} with a configured tracer implementation, or:</li>
+ * <li>Automatically by using the Java <code>ServiceLoader</code> SPI mechanism to load an implementation
+ * from the classpath.</li>
+ * </ol>
+ */
+public final class GlobalTracer implements Tracer {
+    private static final Logger LOGGER = Logger.getLogger(GlobalTracer.class.getName());
+
+    /**
+     * Singleton instance.
+     * <p>
+     * Since we cannot prevent people using {@linkplain #get() GlobalTracer.get()} as a constant,
+     * this guarantees that references obtained before, during or after initialization
+     * all behave as if obtained <em>after</em> initialization once properly initialized.<br>
+     * As a minor additional benefit it makes it harder to circumvent the {@link Tracer} API.
+     */
+    private static final GlobalTracer INSTANCE = new GlobalTracer();
+
+    /**
+     * The resolved {@link Tracer} to delegate the global tracing implementation to.<br>
+     * This can be either an {@link #set(Tracer) explicitly set delegate}
+     * or the automatically resolved Tracer implementation.
+     */
+    private final AtomicReference<Tracer> globalTracer = new AtomicReference<Tracer>();
+
+    private GlobalTracer() {
+    }
+
+    private Tracer lazyTracer() {
+        Tracer instance = globalTracer.get();
+        if (instance == null) {
+            final Tracer singleton = loadSingleSpiImplementation();
+            while (instance == null && singleton != null) { // handle rare race condition
+                globalTracer.compareAndSet(null, singleton);
+                instance = globalTracer.get();
+            }
+            LOGGER.log(Level.INFO, "Using GlobalTracer implementation: {0}.", instance);
+        }
+        LOGGER.log(Level.FINEST, "GlobalTracer: {0}.", instance);
+        return instance;
+    }
+
+    /**
+     * Returns the {@linkplain GlobalTracer} instance.
+     * Upon first use of any tracing method, this tracer lazily determines which actual {@link Tracer}
+     * implementation to use:
+     * <ol type="a">
```

Comment:
these notes make me think the instance will change (but it doesn't).

Also, it isn't just the first use.. it is each use, right? So in other words, the reference returned from this method is constant, but the implementation can change between any calls.

In general, I'd suggest doing a word budget and dropping anything unsurprising or uninteresting. For example, instead of spending words explaining the basics of service loader, ex that it looks at files in META-INF based on the name of the interface, just say "one is looked up from service loader".

you can then spend those words on other things.

Moreover, I'd personally put these details in the type-level javadoc as get() is actually quite simple as it returns a constant. You can summarize much quicker here for those calling it.

Ex.

Returns a constant which forwards all methods to the currently configured tracer.  Until a tracer is explicitly configured, one is looked up from service loader, falling back to noop. A tracer can be re-configured at any time. For example, the tracer used to extract a span may be different than the one that injects it.